<a href="https://colab.research.google.com/github/greypen2006/biophilic-design/blob/main/Clean.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [9]:
from google.colab import files
uploaded = files.upload()



Saving unemployment.raw1.csv to unemployment.raw1.csv


In [4]:
from matplotlib import pyplot as plt
from matplotlib import pylab
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import matplotlib.colors as mcolors
import matplotlib as mpl
from matplotlib import cm
from matplotlib.colors import LinearSegmentedColormap
import matplotlib.patches as mpatches



In [14]:
df=pd.read_csv('unemployment.raw1.csv')
df

,local authority: district / unitary (as of April 2023),mnemonic,Jul 2016-Jun 2017,Conf,Jul 2017-Jun 2018,Conf.1,Jul 2018-Jun 2019,Conf.2,Jul 2019-Jun 2020,Conf.3,Jul 2020-Jun 2021,Conf.4,Jul 2021-Jun 2022,Conf.5,Jul 2022-Jun 2023,Conf.6,Jul 2023-Jun 2024,Conf.7,Jul 2024-Jun 2025,Conf.8
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Brent,E09000005,6.6,1.5,5.9,1.4,4.1,1.4,5.9,1.9,7.6,2.1,5.4,1.8,7.3,3.1,5.5,2.3,8.2,3.0
2,Camden,E09000007,5.3,1.3,5.0,1.4,4.8,1.6,4.0,1.4,5.3,1.7,3.8,1.4,4.0,1.7,6.3,2.2,5.0,2.4
3,Greenwich,E09000011,6.1,1.4,5.5,1.3,4.3,1.3,4.9,1.4,6.6,1.5,5.9,1.5,4.7,1.5,4.1,1.4,6.4,2.3
4,Hackney,E09000012,5.7,1.4,5.7,1.4,5.0,1.5,5.4,1.6,6.1,1.8,4.3,1.5,4.3,1.7,4.4,1.7,5.3,1.8
5,Hammersmith and Fulham,E09000013,5.8,1.4,5.6,1.5,5.8,1.8,4.5,1.5,5.8,1.7,4.7,1.7,3.9,1.7,5.7,2.1,4.3,1.9
6,Islington,E09000019,6.3,1.5,5.0,1.3,5.1,1.6,4.4,1.5,6.2,1.9,3.9,1.4,4.1,1.7,5.9,2.0,4.6,2.1
7,Kensington and Chelsea,E09000020,6.0,1.4,6.2,1.7,5.7,1.8,4.0,1.4,6.0,1.8,4.1,1.5,4.7,2.1,5.7,2.1,5.7,2.6
8,Lambeth,E09000022,5.0,1.3,5.5,1.4,5.6,1.6,5.4,1.6,6.2,1.7,3.5,1.1,3.9,1.5,4.4,1.8,6.4,2.1
9,Southwark,E09000028,6.2,1.4,6.1,1.4,5.0,1.5,5.1,1.5,6.3,1.7,4.3,1.4,4.1,1.5,4.7,1.9,6.5,2.2


In [15]:
clean_df = df.dropna(how='all')
clean_df.to_csv('unemployment_cleaned.csv', index=False)
print("Cleaned DataFrame saved to 'unemployment_cleaned.csv'")

Cleaned DataFrame saved to 'unemployment_cleaned.csv'


In [16]:
clean_df

,local authority: district / unitary (as of April 2023),mnemonic,Jul 2016-Jun 2017,Conf,Jul 2017-Jun 2018,Conf.1,Jul 2018-Jun 2019,Conf.2,Jul 2019-Jun 2020,Conf.3,Jul 2020-Jun 2021,Conf.4,Jul 2021-Jun 2022,Conf.5,Jul 2022-Jun 2023,Conf.6,Jul 2023-Jun 2024,Conf.7,Jul 2024-Jun 2025,Conf.8
1,Brent,E09000005,6.6,1.5,5.9,1.4,4.1,1.4,5.9,1.9,7.6,2.1,5.4,1.8,7.3,3.1,5.5,2.3,8.2,3.0
2,Camden,E09000007,5.3,1.3,5.0,1.4,4.8,1.6,4.0,1.4,5.3,1.7,3.8,1.4,4.0,1.7,6.3,2.2,5.0,2.4
3,Greenwich,E09000011,6.1,1.4,5.5,1.3,4.3,1.3,4.9,1.4,6.6,1.5,5.9,1.5,4.7,1.5,4.1,1.4,6.4,2.3
4,Hackney,E09000012,5.7,1.4,5.7,1.4,5.0,1.5,5.4,1.6,6.1,1.8,4.3,1.5,4.3,1.7,4.4,1.7,5.3,1.8
5,Hammersmith and Fulham,E09000013,5.8,1.4,5.6,1.5,5.8,1.8,4.5,1.5,5.8,1.7,4.7,1.7,3.9,1.7,5.7,2.1,4.3,1.9
6,Islington,E09000019,6.3,1.5,5.0,1.3,5.1,1.6,4.4,1.5,6.2,1.9,3.9,1.4,4.1,1.7,5.9,2.0,4.6,2.1
7,Kensington and Chelsea,E09000020,6.0,1.4,6.2,1.7,5.7,1.8,4.0,1.4,6.0,1.8,4.1,1.5,4.7,2.1,5.7,2.1,5.7,2.6
8,Lambeth,E09000022,5.0,1.3,5.5,1.4,5.6,1.6,5.4,1.6,6.2,1.7,3.5,1.1,3.9,1.5,4.4,1.8,6.4,2.1
9,Southwark,E09000028,6.2,1.4,6.1,1.4,5.0,1.5,5.1,1.5,6.3,1.7,4.3,1.4,4.1,1.5,4.7,1.9,6.5,2.2
10,Tower Hamlets,E09000030,8.1,1.9,7.1,1.8,6.1,1.9,4.7,1.6,6.8,2.0,5.2,1.8,5.2,2.2,5.9,2.0,5.9,2.5


In [17]:
from google.colab import files
files.download('unemployment_cleaned.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [18]:
from google.colab import files
uploaded = files.upload()

Saving MPS Borough Level Crime (Historical).csv to MPS Borough Level Crime (Historical).csv


In [19]:
from google.colab import files
uploaded = files.upload()

Saving MPS Borough Level Crime (most recent 24 months)(1).csv to MPS Borough Level Crime (most recent 24 months)(1).csv


In [21]:
df1=pd.read_csv('MPS Borough Level Crime (Historical).csv')
df1

,MajorText,MinorText,BoroughName,201004,201005,201006,201007,201008,201009,201010,...,202302,202303,202304,202305,202306,202307,202308,202309,202310,202311
0,ARSON AND CRIMINAL DAMAGE,ARSON,Barking and Dagenham,6,5,11,10,6,6,13,...,3,2,2,5,2,5,7,4,4,3
1,ARSON AND CRIMINAL DAMAGE,CRIMINAL DAMAGE,Barking and Dagenham,208,190,218,217,203,161,196,...,91,110,97,108,122,131,114,128,82,88
2,BURGLARY,BURGLARY BUSINESS AND COMMUNITY,Barking and Dagenham,49,58,58,46,46,51,53,...,18,32,36,27,28,25,34,39,26,19
3,BURGLARY,DOMESTIC BURGLARY,Barking and Dagenham,118,102,124,137,153,136,116,...,38,55,48,0,1,1,0,0,0,0
4,BURGLARY,RES BURGLARY OF A HOME,Barking and Dagenham,0,0,0,0,0,0,0,...,0,0,8,44,50,38,65,65,47,49
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1079,VIOLENCE AGAINST THE PERSON,DEATH SERIOUS INJURY ILLEGAL DRIVING,Westminster,0,0,0,0,0,0,0,...,2,0,0,0,0,1,0,0,1,0
1080,VIOLENCE AGAINST THE PERSON,HOMICIDE,Westminster,0,1,0,0,0,0,0,...,0,1,0,0,1,0,0,0,0,0
1081,VIOLENCE AGAINST THE PERSON,STALKING AND HARASSMENT,Westminster,63,74,69,59,42,69,54,...,163,171,165,180,179,194,175,156,167,191
1082,VIOLENCE AGAINST THE PERSON,VIOLENCE WITH INJURY,Westminster,258,280,274,320,271,261,307,...,263,299,290,349,370,373,357,355,333,383


In [23]:
df2=pd.read_csv("MPS Borough Level Crime (most recent 24 months)(1).csv")
df2

,MajorText,MinorText,BoroughName,202312,202401,202402,202403,202404,202405,202406,...,202502,202503,202504,202505,202506,202507,202508,202509,202510,202511
0,ARSON AND CRIMINAL DAMAGE,ARSON,Barking and Dagenham,4,4,5,6,3,8,2,...,8,4,12,10,10,11,10,5,6,8
1,ARSON AND CRIMINAL DAMAGE,CRIMINAL DAMAGE,Barking and Dagenham,125,126,124,134,112,102,106,...,89,103,124,108,124,137,126,100,100,104
2,BURGLARY,BURGLARY BUSINESS AND COMMUNITY,Barking and Dagenham,20,31,21,30,24,28,33,...,14,24,20,27,25,24,23,23,16,21
3,BURGLARY,RES BURGLARY OF A HOME,Barking and Dagenham,44,57,33,45,42,31,31,...,35,58,30,31,35,32,29,31,34,63
4,BURGLARY,RES BURGLARY OF UNCONNECTED BUILDING,Barking and Dagenham,9,10,16,5,10,11,6,...,13,20,11,7,16,5,13,12,22,15
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1051,VIOLENCE AGAINST THE PERSON,DEATH SERIOUS INJURY ILLEGAL DRIVING,Westminster,1,1,0,2,0,0,2,...,0,1,0,0,0,1,0,0,0,2
1052,VIOLENCE AGAINST THE PERSON,HOMICIDE,Westminster,0,0,0,1,1,0,2,...,0,0,0,0,0,0,0,0,0,0
1053,VIOLENCE AGAINST THE PERSON,STALKING AND HARASSMENT,Westminster,157,176,184,125,165,161,128,...,166,178,156,198,171,198,190,183,243,156
1054,VIOLENCE AGAINST THE PERSON,VIOLENCE WITH INJURY,Westminster,379,292,306,265,228,292,267,...,203,237,239,251,304,306,289,271,232,282


In [24]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import random
%matplotlib inline

In [27]:
df3 = pd.concat([df1,df2], ignore_index=True)
df3

,MajorText,MinorText,BoroughName,201004,201005,201006,201007,201008,201009,201010,...,202502,202503,202504,202505,202506,202507,202508,202509,202510,202511
0,ARSON AND CRIMINAL DAMAGE,ARSON,Barking and Dagenham,6.0,5.0,11.0,10.0,6.0,6.0,13.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,ARSON AND CRIMINAL DAMAGE,CRIMINAL DAMAGE,Barking and Dagenham,208.0,190.0,218.0,217.0,203.0,161.0,196.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,BURGLARY,BURGLARY BUSINESS AND COMMUNITY,Barking and Dagenham,49.0,58.0,58.0,46.0,46.0,51.0,53.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,BURGLARY,DOMESTIC BURGLARY,Barking and Dagenham,118.0,102.0,124.0,137.0,153.0,136.0,116.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,BURGLARY,RES BURGLARY OF A HOME,Barking and Dagenham,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2135,VIOLENCE AGAINST THE PERSON,DEATH SERIOUS INJURY ILLEGAL DRIVING,Westminster,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,0.0,1.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,2.0
2136,VIOLENCE AGAINST THE PERSON,HOMICIDE,Westminster,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2137,VIOLENCE AGAINST THE PERSON,STALKING AND HARASSMENT,Westminster,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,166.0,178.0,156.0,198.0,171.0,198.0,190.0,183.0,243.0,156.0
2138,VIOLENCE AGAINST THE PERSON,VIOLENCE WITH INJURY,Westminster,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,203.0,237.0,239.0,251.0,304.0,306.0,289.0,271.0,232.0,282.0


# Task
```python
# Melt df1
id_vars_df1 = ['MajorText', 'MinorText', 'BoroughName']
value_vars_df1 = df1.columns.difference(id_vars_df1).tolist()
df1_melted = pd.melt(df1, id_vars=id_vars_df1, value_vars=value_vars_df1, var_name='YearMonth', value_name='CrimeCount')

# Melt df2
id_vars_df2 = ['MajorText', 'MinorText', 'BoroughName']
value_vars_df2 = df2.columns.difference(id_vars_df2).tolist()
df2_melted = pd.melt(df2, id_vars=id_vars_df2, value_vars=value_vars_df2, var_name='YearMonth', value_name='CrimeCount')

# Concatenate the melted dataframes
combined_df = pd.concat([df1_melted, df2_melted], ignore_index=True)

# Display the first few rows and info of the combined DataFrame
print("Combined DataFrame head:")
print(combined_df.head())
print("\nCombined DataFrame info:")
combined_df.info()
```

## Melt and Concatenate DataFrames

### Subtask:
Transform `df1` and `df2` from wide to long format (year-month columns become rows), then combine them into a single DataFrame.


**Reasoning**:
The first step is to transform `df1` from a wide format to a long format. This involves identifying the ID variables and the value variables (which are the year-month columns) and then using the `melt` function to create new 'YearMonth' and 'CrimeCount' columns.



In [28]:
id_vars = ['MajorText', 'MinorText', 'BoroughName']
value_vars_df1 = [col for col in df1.columns if col not in id_vars]
df1_melted = pd.melt(df1, id_vars=id_vars, value_vars=value_vars_df1, var_name='YearMonth', value_name='CrimeCount')
df1_melted.head()

,MajorText,MinorText,BoroughName,YearMonth,CrimeCount
0,ARSON AND CRIMINAL DAMAGE,ARSON,Barking and Dagenham,201004,6
1,ARSON AND CRIMINAL DAMAGE,CRIMINAL DAMAGE,Barking and Dagenham,201004,208
2,BURGLARY,BURGLARY BUSINESS AND COMMUNITY,Barking and Dagenham,201004,49
3,BURGLARY,DOMESTIC BURGLARY,Barking and Dagenham,201004,118
4,BURGLARY,RES BURGLARY OF A HOME,Barking and Dagenham,201004,0


**Reasoning**:
Next, melt `df2` from wide to long format using the same ID variables, creating 'YearMonth' and 'CrimeCount' columns, and then display the first few rows of the melted DataFrame.



In [29]:
id_vars = ['MajorText', 'MinorText', 'BoroughName']
value_vars_df2 = [col for col in df2.columns if col not in id_vars]
df2_melted = pd.melt(df2, id_vars=id_vars, value_vars=value_vars_df2, var_name='YearMonth', value_name='CrimeCount')
df2_melted.head()

,MajorText,MinorText,BoroughName,YearMonth,CrimeCount
0,ARSON AND CRIMINAL DAMAGE,ARSON,Barking and Dagenham,202312,4
1,ARSON AND CRIMINAL DAMAGE,CRIMINAL DAMAGE,Barking and Dagenham,202312,125
2,BURGLARY,BURGLARY BUSINESS AND COMMUNITY,Barking and Dagenham,202312,20
3,BURGLARY,RES BURGLARY OF A HOME,Barking and Dagenham,202312,44
4,BURGLARY,RES BURGLARY OF UNCONNECTED BUILDING,Barking and Dagenham,202312,9


**Reasoning**:
Now that both `df1` and `df2` have been melted into a long format, the next step is to concatenate these two DataFrames (`df1_melted` and `df2_melted`) into a single DataFrame called `combined_df`.



In [30]:
combined_df = pd.concat([df1_melted, df2_melted], ignore_index=True)
combined_df.head()

,MajorText,MinorText,BoroughName,YearMonth,CrimeCount
0,ARSON AND CRIMINAL DAMAGE,ARSON,Barking and Dagenham,201004,6
1,ARSON AND CRIMINAL DAMAGE,CRIMINAL DAMAGE,Barking and Dagenham,201004,208
2,BURGLARY,BURGLARY BUSINESS AND COMMUNITY,Barking and Dagenham,201004,49
3,BURGLARY,DOMESTIC BURGLARY,Barking and Dagenham,201004,118
4,BURGLARY,RES BURGLARY OF A HOME,Barking and Dagenham,201004,0


**Reasoning**:
To finalize the data merging process and verify the structure and data types of the newly created `combined_df`, the next step is to display a concise summary using the `.info()` method.



In [31]:
combined_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 203120 entries, 0 to 203119
Data columns (total 5 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   MajorText    203120 non-null  object
 1   MinorText    203120 non-null  object
 2   BoroughName  203120 non-null  object
 3   YearMonth    203120 non-null  object
 4   CrimeCount   203120 non-null  int64 
dtypes: int64(1), object(4)
memory usage: 7.7+ MB


## Extract and Filter by Year

### Subtask:
Extract the year from the 'YearMonth' column and filter the combined DataFrame to include only data for the years 2017 to 2024.


**Reasoning**:
The first step is to extract the year from the 'YearMonth' column in `combined_df` by converting it to a string, taking the first four characters, and then converting it to an integer type, as per the instructions.



In [32]:
combined_df['Year'] = combined_df['YearMonth'].astype(str).str[:4].astype(int)
combined_df.head()

,MajorText,MinorText,BoroughName,YearMonth,CrimeCount,Year
0,ARSON AND CRIMINAL DAMAGE,ARSON,Barking and Dagenham,201004,6,2010
1,ARSON AND CRIMINAL DAMAGE,CRIMINAL DAMAGE,Barking and Dagenham,201004,208,2010
2,BURGLARY,BURGLARY BUSINESS AND COMMUNITY,Barking and Dagenham,201004,49,2010
3,BURGLARY,DOMESTIC BURGLARY,Barking and Dagenham,201004,118,2010
4,BURGLARY,RES BURGLARY OF A HOME,Barking and Dagenham,201004,0,2010


**Reasoning**:
Now that the 'Year' column has been extracted, the next step is to filter the `combined_df` to include only rows where the 'Year' is between 2017 and 2024 (inclusive) to fulfill the subtask's requirement.



In [33]:
combined_df_filtered = combined_df[(combined_df['Year'] >= 2017) & (combined_df['Year'] <= 2024)]
combined_df_filtered.head()

,MajorText,MinorText,BoroughName,YearMonth,CrimeCount,Year
87804,ARSON AND CRIMINAL DAMAGE,ARSON,Barking and Dagenham,201701,5,2017
87805,ARSON AND CRIMINAL DAMAGE,CRIMINAL DAMAGE,Barking and Dagenham,201701,125,2017
87806,BURGLARY,BURGLARY BUSINESS AND COMMUNITY,Barking and Dagenham,201701,24,2017
87807,BURGLARY,DOMESTIC BURGLARY,Barking and Dagenham,201701,118,2017
87808,BURGLARY,RES BURGLARY OF A HOME,Barking and Dagenham,201701,0,2017


## Filter by Boroughs

### Subtask:
Filter the DataFrame to include only the specified boroughs: Camden, Islington, Hackney, Tower Hamlets, Greenwich, Southwark, Lambeth, Hammersmith and Fulham, Kensington and Chelsea, Westminster, and Brent.


**Reasoning**:
The first step is to define the list of specified boroughs as `borough_list` to be used for filtering.



In [34]:
borough_list = ['Camden', 'Islington', 'Hackney', 'Tower Hamlets', 'Greenwich', 'Southwark', 'Lambeth', 'Hammersmith and Fulham', 'Kensington and Chelsea', 'Westminster', 'Brent']
print("Borough list created.")

Borough list created.


**Reasoning**:
Now that the `borough_list` has been defined, the next step is to filter the `combined_df_filtered` DataFrame to include only the rows where the 'BoroughName' is present in this list, as per the instructions, and store it in `combined_df_borough_filtered`.



In [35]:
combined_df_borough_filtered = combined_df_filtered[combined_df_filtered['BoroughName'].isin(borough_list)]
combined_df_borough_filtered.head()

,MajorText,MinorText,BoroughName,YearMonth,CrimeCount,Year
87900,ARSON AND CRIMINAL DAMAGE,ARSON,Brent,201701,3,2017
87901,ARSON AND CRIMINAL DAMAGE,CRIMINAL DAMAGE,Brent,201701,155,2017
87902,BURGLARY,BURGLARY BUSINESS AND COMMUNITY,Brent,201701,56,2017
87903,BURGLARY,DOMESTIC BURGLARY,Brent,201701,195,2017
87904,BURGLARY,RES BURGLARY OF A HOME,Brent,201701,0,2017


In [36]:
combined_df_borough_filtered.to_csv('crime_cleaned.csv', index=False)

In [38]:
files.download('crime_cleaned.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Aggregate and Pivot Data

### Subtask:
Aggregate the crime counts by borough, crime type, and year, and then pivot the table to display years as columns, with 'BoroughName' as the index.


## Summary:

### Data Analysis Key Findings

*   **Data Transformation**: The initial dataframes, `df1` and `df2`, were successfully melted from a wide format (with year-month columns) into a long format. These were then concatenated into a single `combined_df` with 203,120 entries and 5 columns (`MajorText`, `MinorText`, `BoroughName`, `YearMonth`, `CrimeCount`). All columns were found to be non-null.
*   **Year Extraction and Filtering**: A 'Year' column was successfully extracted from the 'YearMonth' column. The `combined_df` was then filtered to retain only data for the years 2017 through 2024, resulting in `combined_df_filtered`.
*   **Borough-Specific Filtering**: The data was further filtered to include only 11 specific boroughs (Camden, Islington, Hackney, Tower Hamlets, Greenwich, Southwark, Lambeth, Hammersmith and Fulham, Kensington and Chelsea, Westminster, and Brent), creating `combined_df_borough_filtered`.

### Insights or Next Steps

*   The data is now in a clean, long format and filtered to the relevant time period and geographical areas, making it suitable for direct aggregation and pivoting as per the task requirements.
*   The next step should be to aggregate the `CrimeCount` by `BoroughName`, `CrimeType` (likely `MajorText` or `MinorText`), and `Year`, and then pivot the resulting table to display years as columns.
